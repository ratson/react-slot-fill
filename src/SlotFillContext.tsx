import React from 'react'

const noopRenderCallback = (): React.ReactNode => null

export class SlotFillManager {
  slotsAndFills = new Map()
  subscribers: Array<{ slotId: string; callback: any }> = []

  setFillForSlot = (slotId: string, renderCallback = noopRenderCallback) => {
    const fillForSlot = this.slotsAndFills.get(slotId)

    if (fillForSlot) {
      return
    }

    this.slotsAndFills.set(slotId, renderCallback)
    this.notify(slotId)
  }

  getFillForSlot = (slotId: string) => {
    const fillById = this.slotsAndFills.get(slotId)

    if (!fillById) {
      return noopRenderCallback
    }

    return fillById
  }

  subscribe = (slotId: string, callback: any) => {
    this.subscribers.push({ slotId, callback })
  }

  unsubscribe = (slotId: string, slotIndex: number) => {
    this.subscribers = this.subscribers.filter(
      (subscriber: any, index: number) =>
        subscriber.slotId === slotId && index === slotIndex
    )
  }

  notify = (slotId: string) => {
    this.subscribers.forEach((subscriber: any, index: number) => {
      if (subscriber.slotId !== slotId) {
        return
      }

      subscriber.callback(index)
    })
  }
}

const SlotFillContext = React.createContext(new SlotFillManager())

export default SlotFillContext
