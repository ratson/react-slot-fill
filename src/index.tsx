export { default as Fill } from './Fill'
export { default as Slot } from './Slot'
