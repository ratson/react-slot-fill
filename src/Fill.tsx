import React from 'react'
import SlotFillContext from './SlotFillContext'

export interface FillProps {
  name: string
}

const Fill: React.FC<FillProps> = ({ name, ...props }) => {
  const ctx = React.useContext(SlotFillContext)
  ctx.setFillForSlot(name, () => props.children)

  return null
}

export default Fill
