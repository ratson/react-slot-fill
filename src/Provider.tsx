import React from 'react'
import SlotFillContext, { SlotFillManager } from './SlotFillContext'

const SlotFillProvider: React.FC = ({ children }) => (
  <SlotFillContext.Provider value={new SlotFillManager()}>
    {children}
  </SlotFillContext.Provider>
)

export default SlotFillProvider
