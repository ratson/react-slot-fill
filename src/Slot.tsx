import React from 'react'
import SlotFillContext from './SlotFillContext'

export interface SlotProps {
  name: string
}

const Slot: React.FC<SlotProps> = ({ name, ...props }) => {
  const ctx = React.useContext(SlotFillContext)
  const [slotIndex, setSlotIndex] = React.useState<number>(-1)
  const [, count] = React.useState(0)
  const forceUpdate = React.useCallback(() => count(c => c + 1), [])

  React.useEffect(() => {
    ctx.subscribe(name, (slotIndexNew: number) => {
      console.warn(
        `Slot: Calling suscribe for slotIndex ${slotIndexNew}, where name is ${name}`
      )

      setSlotIndex(slotIndexNew)
      forceUpdate()
    })

    forceUpdate()

    return () => {
      ctx.unsubscribe(name, slotIndex)
    }
  }, [ctx, name, slotIndex, count, forceUpdate])

  const renderCallback = ctx.getFillForSlot(name)
  const children = renderCallback()

  return children ? React.cloneElement(children, props) : null
}

export default Slot
