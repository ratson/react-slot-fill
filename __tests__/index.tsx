import React from 'react'
import { Fill, Slot } from '../src'
import renderer from 'react-test-renderer'

it('can fill before slot', () => {
  const fillComponent = renderer.create(
    <div>
      <Fill name="slot">
        <span>filled</span>
      </Fill>
      <Slot name="slot" />
    </div>
  )

  expect(fillComponent).toMatchSnapshot()
})

it('can fill after slot', () => {
  const fillComponent = renderer.create(
    <div>
      <Slot name="slot2" />
      <Fill name="slot2">
        <span>filled 2</span>
      </Fill>
    </div>
  )

  expect(fillComponent).toMatchSnapshot()
})
