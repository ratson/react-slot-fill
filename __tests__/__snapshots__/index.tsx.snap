// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`can fill after slot 1`] = `
<div>
  <span>
    filled 2
  </span>
</div>
`;

exports[`can fill before slot 1`] = `
<div>
  <span>
    filled
  </span>
</div>
`;
